lektor/admin/static/js/widgets/ToastEditor.jsx

```diff
@@ -0,0 +1,213 @@
+    editor.eventManager.addEventType('mode-tab-richtext')
+    editor.eventManager.addEventType('mode-tab-markdown')
+    editor.eventManager.listen('mode-tab-richtext', () => { markdownbtn.removeClass('active'); richtextbtn.addClass('active'); editor.changeMode('wysiwyg', true) })
+    editor.eventManager.listen('mode-tab-markdown', () => { richtextbtn.removeClass('active'); markdownbtn.addClass('active'); editor.changeMode('markdown', true) })
+    let activebtn = (this.props.type.default_view === 'richtext') ? richtextbtn : markdownbtn
+    activebtn.addClass('active')
+
+    this.recalculateHeight(editor)
+  }
+
+  onImageUpload (file, cb, source) {
+    // do image upload of file here, call cb once done
+    cb(file.name) // leave second param null to use user-specified text
+  }
+
+  recalculateHeight (editor) {
+    let currentEditor = editor.getCurrentModeEditor()
+    let editorHeight
+    try {
+      // markdown
+      let editorEl = currentEditor.editorContainerEl.getElementsByClassName('CodeMirror-sizer')[0]
+      editorHeight = this.outerHeight(editorEl)
+    } catch (e) {
+      // wysiwyg
+
+      // get height of all children in editor
+      let editorEl = currentEditor.$editorContainerEl[0].firstChild
+      let editorChildren = Array.from(editorEl.children)
+      let editorChildrenHeights = editorChildren.map(el => this.outerHeight(el))
+      editorHeight = editorChildrenHeights.reduce((a, b) => a + b)
+
+      // add on padding/border/margin of editor element
+      let marginTop = parseInt(window.getComputedStyle(editorEl).marginTop)
+      let marginBottom = parseInt(window.getComputedStyle(editorEl).marginBottom)
+      let paddingTop = parseInt(window.getComputedStyle(editorEl).paddingTop)
+      let paddingBottom = parseInt(window.getComputedStyle(editorEl).paddingBottom)
+      let borderTop = parseInt(window.getComputedStyle(editorEl).borderTopWidth)
+      let borderBottom = parseInt(window.getComputedStyle(editorEl).borderBottomWidth)
+      editorHeight += (marginTop + marginBottom + paddingTop + paddingBottom + borderTop + borderBottom)
+    }
+
+    let totalHeight = editorHeight + 31 + 31
+    let totalHeightStr = totalHeight + 'px'
+    editor.height(totalHeightStr)
+  }
+
+  onChange () {
+    if (!this.state.editor) {
+      return
+    }
+
+    const markdown = this.state.editor.getMarkdown()
+    if (this.count === 0 && markdown === '') {
+      this.count++
+      return
+    }
+
+    // recalculate height - delay is required to get most up to date height
+    setTimeout(function () {
+      this.recalculateHeight(this.state.editor)
+    }.bind(this), 10)
+
+    // send markdown up
+    let contentChanged = markdown !== this.content
+    if (this.props.onChange && this.content !== null && contentChanged) {
+      this.content = markdown
+      this.props.onChange(markdown)
+    }
+  }
+
+  // focus and blur are used to implement the border-blur found on other widgets
+  focus () {
+    this.setState({
```

Comment:
Is it possible to use Lektor's existing CSS here? I think it's better to fall back on that so you don't have to change it in two places. The same comment applies to `blur()`.

Reply:
Fixed in 8671ab8.
It's not possible to just use existing CSS so I've had to add some extra rules in main.less.